src/libostree/ostree-repo-libarchive.c

```diff
@@ -45,269 +47,685 @@ propagate_libarchive_error (GError      **error,
+  g_autofree guchar *csum_raw = NULL;
+  g_autofree char *csum = NULL;
+
+  if (name[0] == '\0') /* root? */
+    dir = g_object_ref (parent);
+  else if (ostree_mutable_tree_lookup (parent, name, NULL, &dir, error))
     {
-    case G_FILE_TYPE_REGULAR:
-      archive_stream = _ostree_libarchive_input_stream_new (a);
-      break;
-    case G_FILE_TYPE_SYMBOLIC_LINK:
-      break;
-    default:
-      if (opts->ignore_unsupported_content)
-        {
-          ret = TRUE;
-          goto out;
-        }
-      else
+      if (error_if_exist)
         {
           g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
-                       "Unable to import non-regular/non-symlink file '%s'",
-                       archive_entry_pathname (entry));
-          goto out;
+                       "Directory \"%s\" already exists", name);
+          return FALSE;
         }
     }
-  
-  if (!ostree_raw_file_to_content_stream (archive_stream, file_info, NULL,
-                                          &file_object_input, &length, cancellable, error))
-    goto out;
-  
-  if (!ostree_repo_write_content (self, NULL, file_object_input, length, out_csum,
+
+  if (dir == NULL)
+    {
+      if (!g_error_matches (*error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
+        return FALSE;
+
+      g_clear_error (error);
+
+      if (!ostree_mutable_tree_ensure_dir (parent, name, &dir, error))
+        return FALSE;
+    }
+
+  if (!_ostree_repo_write_directory_meta (repo, file_info, xattrs,
+                                          &csum_raw, cancellable, error))
+    return FALSE;
+
+  csum = ostree_checksum_from_bytes (csum_raw);
+
+  ostree_mutable_tree_set_metadata_checksum (dir, csum);
+
+  if (out_dir)
+    *out_dir = g_steal_pointer (&dir);
+
+  return TRUE;
+}
+
+typedef struct {
+  OstreeRepo                     *repo;
+  OstreeRepoImportArchiveOptions *opts;
+  OstreeMutableTree              *root;
+  struct archive                 *archive;
+  struct archive_entry           *entry;
+  GHashTable                     *deferred_hardlinks;
+  OstreeRepoCommitModifier       *modifier;
+} OstreeRepoArchiveImportContext;
+
+typedef struct {
+  OstreeMutableTree  *parent;
+  char               *path;
+  guint64             size;
+} DeferredHardlink;
+
+static inline char*
+aic_get_final_path (OstreeRepoArchiveImportContext *ctx,
+                    const char  *path)
+{
+  if (ctx->opts->use_ostree_convention)
+    return path_relative_ostree (path);
+  return g_strdup (path_relative (path));
+}
+
+static inline char*
+aic_get_final_entry_pathname (OstreeRepoArchiveImportContext *ctx)
+{
+  const char *pathname = archive_entry_pathname (ctx->entry);
+  g_autofree char *final = aic_get_final_path (ctx, pathname);
+
+  /* squash trailing slashes some archives append to dir names */
+  if (g_str_has_suffix (final, "/"))
+    {
+      g_autofree char *tmp = final;
+      final = g_strndup (final, strlen (final) - 1);
```

Comment:
Also need to handle multiple trailing `/` here?  Would be a bit more efficient too to get the end pointer, then walk backwards and replace them with `NUL` in place.